### PR TITLE
SSP2.x

### PR DESCRIPTION
**Pull request by clmcavaney**
Have made some changes to allow the module to run with SimpleSAML v2.x - based on the [upgrade to 2.0 documentation](https://simplesamlphp.org/docs/stable/simplesamlphp-upgrade-notes-2.0.html).

Also added the ability for a validation timeout to be specified (defaults to 60 minutes).

### CODE REVIEW

File: public/not_configured.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+use SimpleSAML\Configuration;
+use SimpleSAML\XHTML\Template;
+
+$globalConfig = Configuration::getInstance();
+$t = new Template($globalConfig, 'simpletotp:not_configured.twig');
+//$t->show();
```
Comment (aidan-) [inline public/not_configured.php:8]: nit: let's remove this commented out code

File: public/authenticate.php [hunk 2]
```diff
@@ -17,12 +26,14 @@
 
 $id = $_REQUEST['StateId'];
 
-$sid = SimpleSAML_Utilities::parseStateID($id);
+//$sid = SimpleSAML_Utilities::parseStateID($id);
+$sid = State::parseStateID($id);
 if (!is_null($sid['url'])) {
-    SimpleSAML_Utilities::checkURLAllowed($sid['url']);
+	$httpUtils = new Utils\HTTP();
```
Comment (aidan-) [inline public/authenticate.php:32]:
```
    $httpUtils = new Utils\HTTP();
```

File: public/authenticate.php [hunk 3]
```diff
@@ -17,12 +26,14 @@
 
 $id = $_REQUEST['StateId'];
 
-$sid = SimpleSAML_Utilities::parseStateID($id);
+//$sid = SimpleSAML_Utilities::parseStateID($id);
```
Comment (aidan-) [inline public/authenticate.php:29]:
```suggestion
```

File: public/authenticate.php [hunk 4]
```diff
@@ -32,24 +43,24 @@
...
 
 // populate values for template
-$t = new SimpleSAML_XHTML_Template($globalConfig, 'simpletotp:authenticate.php');
+$t = new Template($globalConfig, 'simpletotp:authenticate.twig');
 $t->data['formData'] = array('StateId' => $id);
-$t->data['formPost'] = SimpleSAML\Module::getModuleURL('simpletotp/authenticate.php');
+$t->data['formPost'] = Module::getModuleURL('simpletotp/authenticate.php');
 $t->data['userError'] = $displayed_error;
-$t->show();
+//$t->show();
```
Comment (aidan-) [inline public/authenticate.php:62]:
```suggestion
```

File: src/Auth/Process/mfa.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+        $attributes =& $state['Attributes'];
+
+        // check for secret_attr coming from user store and make sure it is not empty
+        if (array_key_exists($this->secret_attr, $attributes) && !empty($attributes[$this->secret_attr])) {
+            $this->secret_val = $attributes[$this->secret_attr][0];
+        }
+
+        if ($this->secret_val === NULL && $this->enforce_mfa === true) {
+            # 2f is enforced and user does not have it configured..
```
Comment (aidan-) [inline src/Auth/Process/mfa.php:127]:
```suggestion
            # mfa is enforced and user does not have it configured..
```

File: src/Auth/Process/mfa.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            if ($this->not_configured_url !== NULL) {
+                $httpUtils = new Utils\HTTP();
+                $httpUtils->redirectUntrustedURL($this->not_configured_url);
+            } else {
+                $httpUtils = new Utils\HTTP();
+                $httpUtils->redirectTrustedURL(Module::getModuleURL('simpletotp/not_configured.php'));
+            }
+
+        } elseif ($this->secret_val === NULL && $this->enforce_mfa === false) {
+            Logger::debug('User with ID "' . $attributes['uid'][0] . '" does not have 2f configured but SP does not require it. Continue.');
```
Comment (aidan-) [inline src/Auth/Process/mfa.php:140]:
```suggestion
            Logger::debug('User with ID "' . $attributes['uid'][0] . '" does not have mfa configured but SP does not require it. Continue.');
```

File: src/Auth/Process/mfa.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+                'lastverified',
+                $now,
+                Session::DATA_TIMEOUT_SESSION_END
+            );
+            if ( $lastverified === NULL ) {
+                $reason = 'new session';
+            } else {
+                $reason = ($now - $lastverified) . 's ago';
+            }
+            Logger::info('MFA: verification required.  New session or last verified more than an hour ago - ' . $reason);
```
Comment (aidan-) [inline src/Auth/Process/mfa.php:175]:
```suggestion
            Logger::info('MFA: verification required.  New session or last verified more than ' . $this->validation_timeout . ' minutes ago - ' . $reason);
```

File: src/Auth/Process/mfa.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        } else {
+            // nothing more to do here
+            Logger::info('MFA: already verified in the last hour - ' . ($now - $lastverified) . 's ago');
```
Comment (aidan-) [inline src/Auth/Process/mfa.php:178]:
```suggestion
            Logger::info('MFA: already verified in the last ' . $this->validation_timeout . ' minutes - ' . ($now - $lastverified) . 's ago');
```

File: templates/authenticate.twig [hunk 9]
```diff
@@ -0,0 +1,44 @@
+{#
+ * Template form for submitting 2fa codes
```
Comment (aidan-) [inline templates/authenticate.twig:2]:
```suggestion
 * Template form for submitting mfa codes
```

File: templates/authenticate.twig [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+{% set pagetitle = 'TOTP'|trans %}
+{% extends '@core/base.twig' %}
+
+{% block preload %}
+<link rel="stylesheet" type="test/css" href="/{{ baseurlpath }}/module.php/simpletotp/style.css" />
+{% endblock %}
+
+{% block content %}
+
+<h1>Two-Factor Authentication Required</h1>
```
Comment (aidan-) [inline templates/authenticate.twig:13]:
```suggestion
<h1>Multi-Factor Authentication Required</h1>
```

File: templates/authenticate.twig [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+<p>You are required to enter your current two-factor authentication token using your configured device to continue. <br />
```
Comment (aidan-) [inline templates/authenticate.twig:14]:
```suggestion
<p>You are required to enter your current multi-factor authentication token using your configured device to continue. <br />
```

File: templates/authenticate.twig [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+If you are unsure what this means, or have your lost your two-factor authentication device, you will need to contact your IT helpdesk for assistance.</p>
```
Comment (aidan-) [inline templates/authenticate.twig:15]:
```suggestion
If you are unsure what this means, or have your lost your multi-factor authentication device, you will need to contact your IT helpdesk for assistance.</p>
```

File: templates/not_configured.twig [hunk 13]
```diff
@@ -0,0 +1,19 @@
+{#
+ * Template for error message users receive when they do not have a TOTP
+ *  value when 2fa is enforced.
```
Comment (aidan-) [inline templates/not_configured.twig:3]:
```suggestion
 *  value when mfa is enforced.
```

File: templates/not_configured.twig [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+{% set pagetitle = 'TOTP'|trans %}
+{% extends '@core/base.twig' %}
+
+{% block preload %}
+<link rel="stylesheet" type="test/css" href="/{{ baseurlpath }}/module.php/simpletotp/style.css" />
+{% endblock %}
+
+{% block content %}
+
+<h1>Two-Factor Authentication Required</h1>
```
Comment (aidan-) [inline templates/not_configured.twig:15]:
```suggestion
<h1>Multi-Factor Authentication Required</h1>
```

File: README.md [hunk 15]
```diff
@@ -28,10 +28,11 @@ After the module has been called, the attribute will be moved out of the user at
 Placed in either config.php's authproc or in the appropriate metadata entity:
 ```php
 10 => array(
-	'class' => 'simpletotp:2fa',
+	'class' => 'simpletotp:mfa',
 	'secret_attr' => 'ga_secret', //default
-	'enforce_2fa' => false, //default
+	'enforce_mfa' => false, //default
 	'not_configured_url' => NULL,  //default
+	'validation_timeout' => <int>, //default 60 (units: minutes) - optional timeout value for re-validation
```
Comment (aidan-) [inline README.md:35]:
```suggestion
	'validation_timeout' => 60, //default, (units: minutes) - number of minutes to cache a successful user mfa validation```

**Review by aidan- — COMMENTED** [review]
Thank you for putting together this PR!  I don't personally use this module anymore so I do not have an environment ready in which I can validate this, but from my very rusty PHP, it looks good!

I've just put some quick nits in to remove mentions of '2fa' in favor of 'mfa' to keep things consistent, and some suggestions for log messages/comments.

File: public/not_configured.php (hunk 1 above)
Comment (clmcavaney) [reply public/not_configured.php:8]: oops - I forgot to remove all of those.  Will work through them.

**Comment by clmcavaney** [comment]
All changes made and tweaked some other bits.
Thanks for the review on that @aidan-, I missed some comments there.

**Pull request merged**